src/hotspot/cpu/aarch64/vm_version_aarch64.cpp

```diff
@@ -452,29 +452,23 @@ void VM_Version::initialize() {
 
   if (UseBranchProtection == nullptr || strcmp(UseBranchProtection, "none") == 0) {
     _rop_protection = false;
-  } else if (strcmp(UseBranchProtection, "standard") == 0) {
+  } else if (strcmp(UseBranchProtection, "standard") == 0 ||
+             strcmp(UseBranchProtection, "pac-ret") == 0) {
     _rop_protection = false;
     // Enable PAC if this code has been built with branch-protection, the CPU/OS
     // supports it, and incompatible preview features aren't enabled.
-#ifdef __ARM_FEATURE_PAC_DEFAULT
-    if (VM_Version::supports_paca() && !Arguments::enable_preview()) {
-      _rop_protection = true;
-    }
-#endif
-  } else if (strcmp(UseBranchProtection, "pac-ret") == 0) {
-    _rop_protection = true;
 #ifdef __ARM_FEATURE_PAC_DEFAULT
     if (!VM_Version::supports_paca()) {
-      warning("ROP-protection specified, but not supported on this CPU.");
       // Disable PAC to prevent illegal instruction crashes.
-      _rop_protection = false;
+      warning("ROP-protection specified, but not supported on this CPU. Disabling ROP-protection.");
     } else if (Arguments::enable_preview()) {
       // Not currently compatible with continuation freeze/thaw.
-      warning("PAC-RET is incompatible with virtual threads preview feature.");
-      _rop_protection = false;
+      warning("ROP-protection is incompatible with virtual threads preview feature. Disabling ROP-protection.");
```

Comment:
```suggestion
      _rop_protection = false;
      warning("ROP-protection is incompatible with virtual threads preview feature. Disabling ROP-protection.");
```

Reply:
I presume you didn't intend to change the logic here.

Reply:
Virtual threads are proposed to be a permanent feature in JDK 21 so I assume predicating this on Arguments:enable_preview check no longer make sense.

Reply:
> Virtual threads are proposed to be a permanent feature in JDK 21 so I assume predicating this on Arguments:enable_preview check no longer make sense.

I knew this issue.
Initially I want to make this patch as small as possible, i.e. only focusing on the SIGILL issue.

> Oh, I see. So it really only makes sense either to make ROP protection work in vthreads, or just disable ROP protection in JDK 21.

It may take some more time for me to "make ROP protection work in vthreads", i.e. resolving JDK-8287325. 
Compared to the zero modifier solution (See https://github.com/openjdk/jdk/pull/13322), I personally think relative sp solution is better, but we still have many jtreg failures to fix. I'm not sure if we can make it ready in JDK21 release cycle.

Perhaps, we can firstly disable ROP if `VMContinuations` is on. I mean, change `Arguments::enable_preview()` to `VMContinuations`.
In this way, ROP protection is enabled only with VM options `-XX:UseBranchProtection=standard|pac-ret -XX:+UnlockExperimentalVMOptions -XX:-VMContinuations` on hardware with the support of PAC feature.

Reply:
> Perhaps, we can firstly disable ROP if `VMContinuations` is on. I mean, change `Arguments::enable_preview()` to `VMContinuations`.
> In this way, ROP protection is enabled only with VM options `-XX:UseBranchProtection=standard|pac-ret -XX:+UnlockExperimentalVMOptions -XX:-VMContinuations` on hardware with the support of PAC feature.

That's probably the best we can do for 21.